Extend default log pattern on server-side to include error information

maryam-saeidi opened the pull request:
## Release Notes
Kibana logging's pattern layout, used by default for the console appender, will now use a new default pattern layout `[%date][%level][%logger] %message %error`. This will include the error name and stack trace if these were included in the log entry. To opt out of this behavior users can omit the `%error` placeholder from their log pattern config in kibana.yml e.g.:
```
logging:
  appenders:
    console:
      type: console
      layout:
        type: pattern
        pattern: "[%date][%level][%logger] %message"
```

## Summary

Previously, when we pass the error in meta, the information related to stacktrace and error message was not available in console. This PR changed the default pattern to also include error information if it is provided in meta (similar to the way that the logging happens when error is directly passed to logger.error).

New pattern: (added `%error` at the end)
```
[%date][%level][%logger] %message %error
```

Here you can see the difference:

Logger:

```
server.logger.error(
        `Unable to create Synthetics monitor ${monitorWithNamespace[ConfigKey.NAME]}`,
        { error: e }
      );
```

#### Before

![image](https://github.com/user-attachments/assets/4f3ff751-84d5-4b5b-b6a9-d49f868a9606)

#### After

![image](https://github.com/user-attachments/assets/e22b8e45-1b0a-4d8c-b51d-5dfb3938da4f)


### Alternative
We could also change the MetaConversion and include this information, but we might have additional meta information which I am not sure if it is OK to be logged by default. Let me know if you prefer changing MetaConversion instead of adding a new error conversion.

<details>
<summary>Code changes for MetaConversion</summary>

```
function isError(x: any): x is Error {
  return x instanceof Error;
}

export const MetaConversion: Conversion = {
  pattern: /%meta/g,
  convert(record: LogRecord) {
    if (!record.meta) {
      return '';
    }
    const { error, ...rest } = record.meta;
    const metaString = Object.keys(rest).length !== 0 ? JSON.stringify(rest) : '';
    let errorString = '';

    if (isError(record.meta?.error)) {
      errorString = record.meta?.error.stack || '';
    }

    return [metaString, errorString].filter(Boolean).join(' ');
  },
};
```
</details>

Here is how adjusting meta will look like in this case:

![image](https://github.com/user-attachments/assets/d7dce9bc-7147-472d-b434-373322f41bbf)

<!--ONMERGE {"backportTargets":["8.19"]} ONMERGE-->

maryam-saeidi committed:
Extend default log pattern on server-side

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/298239)
* Commit: 2d613095b2a502161b97d829b9058824d36ccd55

### Failed CI Steps
* [Jest Tests #5](https://buildkite.com/elastic/kibana-pull-request/builds/298239#01969175-9f84-4750-990d-5ef03d4d4323)
* [Jest Tests #5](https://buildkite.com/elastic/kibana-pull-request/builds/298239#019691a1-2185-4987-b52f-983aa7ca0b30)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/298239#01969175-9f84-4750-990d-5ef03d4d4323) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/298239/jobs/01969175-9f84-4750-990d-5ef03d4d4323/artifacts/019691a0-f3d9-4f1d-9bd1-6d84e12a2bfa) Jest Tests #<span></span>5 / `format()` correctly formats record with full pattern.
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/298239#019691a1-2185-4987-b52f-983aa7ca0b30) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/298239/jobs/019691a1-2185-4987-b52f-983aa7ca0b30/artifacts/019691a5-d0fb-40be-a1cd-1582f0833efd) Jest Tests #<span></span>5 / `format()` correctly formats record with full pattern.
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/298239#01969175-9f84-4750-990d-5ef03d4d4323) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/298239/jobs/01969175-9f84-4750-990d-5ef03d4d4323/artifacts/019691a0-f3d8-489b-85a3-76afc46353f7) Jest Tests #<span></span>5 / `format()` correctly formats record with highlighting.
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/298239#019691a1-2185-4987-b52f-983aa7ca0b30) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/298239/jobs/019691a1-2185-4987-b52f-983aa7ca0b30/artifacts/019691a5-d0fc-4045-8bf6-998280f6e4d5) Jest Tests #<span></span>5 / `format()` correctly formats record with highlighting.
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/298239#01969175-9f84-4750-990d-5ef03d4d4323) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/298239/jobs/01969175-9f84-4750-990d-5ef03d4d4323/artifacts/019691a0-f3d8-4851-bfb7-bba4688f4e39) Jest Tests #<span></span>5 / appends records via multiple appenders.
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/298239#019691a1-2185-4987-b52f-983aa7ca0b30) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/298239/jobs/019691a1-2185-4987-b52f-983aa7ca0b30/artifacts/019691a5-d0fd-4fc0-a5c1-5682838e8e45) Jest Tests #<span></span>5 / appends records via multiple appenders.
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/298239#01969175-9f84-4750-990d-5ef03d4d4323) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/298239/jobs/01969175-9f84-4750-990d-5ef03d4d4323/artifacts/019691a0-f3d8-4e99-8446-d105d2e19078) Jest Tests #<span></span>5 / format timestamp uses ISO8601_TZ as default
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/298239#019691a1-2185-4987-b52f-983aa7ca0b30) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/298239/jobs/019691a1-2185-4987-b52f-983aa7ca0b30/artifacts/019691a5-d0fc-46b7-b799-140813852cd7) Jest Tests #<span></span>5 / format timestamp uses ISO8601_TZ as default
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/298239#01969175-9f84-4750-990d-5ef03d4d4323) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/298239/jobs/01969175-9f84-4750-990d-5ef03d4d4323/artifacts/019691a0-f3da-413e-a40f-05d51f3ab3a2) Jest Tests #<span></span>5 / uses `root` logger if context name is not specified.
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/298239#019691a1-2185-4987-b52f-983aa7ca0b30) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/298239/jobs/019691a1-2185-4987-b52f-983aa7ca0b30/artifacts/019691a5-d0fc-4ffe-a685-e65b51e129f5) Jest Tests #<span></span>5 / uses `root` logger if context name is not specified.

cc @maryam-saeidi

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/298239","number":298239,"commit":"2d613095b2a502161b97d829b9058824d36ccd55"}],"number":298239}
buildkite-pr-comment-->

rudolf commented:
> We could also change the MetaConversion and include this information, but we might have additional meta information which I am not sure if it is OK to be logged by default. 

Yeah logging meta by default would be quite a big change suddenly generating a lot more volume than what users using the pattern layout would be used to. So I think adding `%error` would be preferable.

We probably want to include the errorConversion and change the default pattern layout for the browser too. https://github.com/elastic/kibana/blob/main/src/core/packages/logging/browser-internal/src/layouts/pattern_layout.ts

I noticed there's three DEFAULT_PATTERN constants which is probably unnecessary. We can probably remove the following two: 
 - https://github.com/elastic/kibana/blob/main/src/core/packages/logging/browser-internal/src/layouts/pattern_layout.ts#L20
 - https://github.com/elastic/kibana/blob/main/src/core/packages/logging/server-internal/src/layouts/pattern_layout.ts#L24

And just keep: 
https://github.com/elastic/kibana/blob/main/src/core/packages/logging/common-internal/src/layouts/pattern_layout.ts#L23

Since the common `BasePatternLayout` already sets the default pattern to DEFAULT_PATTERN, I think we can just construct the PatternLayout on server and browser without setting a default for the `pattern` param. E.g. in https://github.com/elastic/kibana/blob/main/src/core/packages/logging/server-internal/src/layouts/pattern_layout.ts#L56
```
 constructor(pattern?: string, highlight: boolean = false) {
 ```

We'll need to update docs in:
https://github.com/elastic/kibana/blob/main/src/core/server/docs/kib_core_logging.mdx?plain=1#L132
https://github.com/elastic/kibana/blob/main/docs/reference/configuration-reference/logging-settings.md?plain=1#L32

maryam-saeidi committed:
Fix tests by trimming the error message

maryam-saeidi committed:
Change defailt log pattern in browser

maryam-saeidi committed:
Remove extra DEFAULT_PATTERNs

maryam-saeidi committed:
Adjust error tests by including meta.error

maryam-saeidi committed:
Merge branch 'main' into extend-server-side-log-pattern

maryam-saeidi commented:
- [ ] Include the errorConversion and change the default pattern layout for the browser too
- [ ] Removing extra DEFAULT_PATTRNs
- [ ] Only setting DEFAULT_PATTERN in the common implementation
- [ ] Update doc

maryam-saeidi committed:
Update docs

maryam-saeidi commented:
@rudolf I updated the PR based on your comment. For the two docs that you mentioned, I just changed the pattern. Let me know if something else needs to be changed as well.

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/299336)
* Commit: c634dfec59600525222c41d33f58214acab9cd39

### Failed CI Steps
* [Build and Validate Next Docs](https://buildkite.com/elastic/kibana-pull-request/builds/299336#0196aa3a-9ad8-4e0e-924e-89526033d607)

### History
* :broken_heart: [Build #298239](https://buildkite.com/elastic/kibana-pull-request/builds/298239) failed 2d613095b2a502161b97d829b9058824d36ccd55

cc @maryam-saeidi

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/299336","number":299336,"commit":"c634dfec59600525222c41d33f58214acab9cd39"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/298239","number":298239,"commit":"2d613095b2a502161b97d829b9058824d36ccd55"}],"number":299336}
buildkite-pr-comment-->

maryam-saeidi committed:
fix test

maryam-saeidi committed:
Merge branch 'main' into extend-server-side-log-pattern

maryam-saeidi committed:
Temporary commit

maryam-saeidi committed:
Revert "Temporary commit"

This reverts commit 71eb0f053e0c205bc168f0fd2cb7fe58f1feb3bb.

maryam-saeidi committed:
Merge branch 'main' into extend-server-side-log-pattern

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/299646)
* Commit: 79ab09358def24c1080d7e09ebe22d84707c7696

### Failed CI Steps
* [Build and Validate Next Docs](https://buildkite.com/elastic/kibana-pull-request/builds/299646#0196af26-1cf6-485f-ba44-0b9328581776)

### History
* :broken_heart: [Build #299458](https://buildkite.com/elastic/kibana-pull-request/builds/299458) failed 15c697b396966356b9818b3ae292f34ad7a93b8a
* :broken_heart: [Build #299336](https://buildkite.com/elastic/kibana-pull-request/builds/299336) failed c634dfec59600525222c41d33f58214acab9cd39
* :broken_heart: [Build #298239](https://buildkite.com/elastic/kibana-pull-request/builds/298239) failed 2d613095b2a502161b97d829b9058824d36ccd55

cc @maryam-saeidi

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/299646","number":299646,"commit":"79ab09358def24c1080d7e09ebe22d84707c7696"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/299458","number":299458,"commit":"15c697b396966356b9818b3ae292f34ad7a93b8a"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/299336","number":299336,"commit":"c634dfec59600525222c41d33f58214acab9cd39"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/298239","number":298239,"commit":"2d613095b2a502161b97d829b9058824d36ccd55"}],"number":299646}
buildkite-pr-comment-->

maryam-saeidi committed:
Merge branch 'main' into extend-server-side-log-pattern

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/300394)
* Commit: ba74dfbd55742cc8c5f275cc560f667fbaa0861a

### Failed CI Steps
* [Build and Validate Next Docs](https://buildkite.com/elastic/kibana-pull-request/builds/300394#0196c8ff-1ee0-4a7d-971b-131de8b4b6f3)

### History
* :broken_heart: [Build #299646](https://buildkite.com/elastic/kibana-pull-request/builds/299646) failed 79ab09358def24c1080d7e09ebe22d84707c7696
* :broken_heart: [Build #299458](https://buildkite.com/elastic/kibana-pull-request/builds/299458) failed 15c697b396966356b9818b3ae292f34ad7a93b8a
* :broken_heart: [Build #299336](https://buildkite.com/elastic/kibana-pull-request/builds/299336) failed c634dfec59600525222c41d33f58214acab9cd39
* :broken_heart: [Build #298239](https://buildkite.com/elastic/kibana-pull-request/builds/298239) failed 2d613095b2a502161b97d829b9058824d36ccd55

cc @maryam-saeidi

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/300394","number":300394,"commit":"ba74dfbd55742cc8c5f275cc560f667fbaa0861a"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/299646","number":299646,"commit":"79ab09358def24c1080d7e09ebe22d84707c7696"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/299458","number":299458,"commit":"15c697b396966356b9818b3ae292f34ad7a93b8a"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/299336","number":299336,"commit":"c634dfec59600525222c41d33f58214acab9cd39"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/298239","number":298239,"commit":"2d613095b2a502161b97d829b9058824d36ccd55"}],"number":300394}
buildkite-pr-comment-->

rudolf reviewed (approved):
Looks good, thanks 🥳

maryam-saeidi committed:
Merge branch 'main' into extend-server-side-log-pattern

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/301590)
* Commit: a4a751373f05c17c2f9fb5248c88d3ad251ebb84

### Failed CI Steps
* [FTR Configs #11](https://buildkite.com/elastic/kibana-pull-request/builds/301590#0196f716-752b-4cff-ab19-bb5eb17896aa)
* [FTR Configs #11](https://buildkite.com/elastic/kibana-pull-request/builds/301590#0196f736-6271-43f0-bb82-5f1cfb42ed0d)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/301590#0196f716-752b-4cff-ab19-bb5eb17896aa) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/301590/jobs/0196f716-752b-4cff-ab19-bb5eb17896aa/artifacts/0196f736-4f81-46e0-9e75-8d0d56e77e25) FTR Configs #<span></span>11 / Cloud Security Posture Vulnerabilities Page - Grouping SearchBar add filter
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/301590#0196f736-6271-43f0-bb82-5f1cfb42ed0d) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/301590/jobs/0196f736-6271-43f0-bb82-5f1cfb42ed0d/artifacts/0196f74d-20a0-4aea-8492-60c6e801712b) FTR Configs #<span></span>11 / Cloud Security Posture Vulnerabilities Page - Grouping SearchBar add filter

### History
* :broken_heart: [Build #300394](https://buildkite.com/elastic/kibana-pull-request/builds/300394) failed ba74dfbd55742cc8c5f275cc560f667fbaa0861a
* :broken_heart: [Build #299646](https://buildkite.com/elastic/kibana-pull-request/builds/299646) failed 79ab09358def24c1080d7e09ebe22d84707c7696
* :broken_heart: [Build #299458](https://buildkite.com/elastic/kibana-pull-request/builds/299458) failed 15c697b396966356b9818b3ae292f34ad7a93b8a
* :broken_heart: [Build #299336](https://buildkite.com/elastic/kibana-pull-request/builds/299336) failed c634dfec59600525222c41d33f58214acab9cd39
* :broken_heart: [Build #298239](https://buildkite.com/elastic/kibana-pull-request/builds/298239) failed 2d613095b2a502161b97d829b9058824d36ccd55

cc @maryam-saeidi

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/301590","number":301590,"commit":"a4a751373f05c17c2f9fb5248c88d3ad251ebb84"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/300394","number":300394,"commit":"ba74dfbd55742cc8c5f275cc560f667fbaa0861a"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/299646","number":299646,"commit":"79ab09358def24c1080d7e09ebe22d84707c7696"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/299458","number":299458,"commit":"15c697b396966356b9818b3ae292f34ad7a93b8a"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/299336","number":299336,"commit":"c634dfec59600525222c41d33f58214acab9cd39"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/298239","number":298239,"commit":"2d613095b2a502161b97d829b9058824d36ccd55"}],"number":301590}
buildkite-pr-comment-->

maryam-saeidi committed:
Merge branch 'main' into extend-server-side-log-pattern

kibanamachine commented:
Starting backport for target branches: 8.19

https://github.com/elastic/kibana/actions/runs/15189967763

kibanamachine commented:
## 💔 All backports failed

| Status | Branch | Result |
|:------:|:------:|:------|
|❌|8.19|**Backport failed because of merge conflicts**<br><br>You might need to backport the following PRs to 8.19:<br> - [Upgrade Puppeteer to v24.8.1 (#220310)](https://github.com/elastic/kibana/pull/220310)<br> - [[ci] Increase disk for capture_oas_snapshot (#221267)](https://github.com/elastic/kibana/pull/221267)<br> - [[Synthetics] Added task to sync global parameters when they are added, edited or deleted (#216015)](https://github.com/elastic/kibana/pull/216015)<br> - [[Security Solution] [Onboarding] a11y: Fix incorrect navigation after Save on AI assistant modal dialog (#219935)](https://github.com/elastic/kibana/pull/219935)|

### Manual backport
To create the backport manually run:
```
node scripts/backport --pr 219940
```

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 9.6.6 -->

maryam-saeidi commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.19|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/221339">](https://github.com/elastic/kibana/pull/221339)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 10.0.0 -->

florent-leborgne commented:
Hi @maryam-saeidi.  Can you add the correct team or feature label to this PR so it gets pulled into the correct release notes? Thank you!

maryam-saeidi commented:
@florent-leborgne I've added "Feature:Logging" label, please let me know if something else is needed :)
